core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala

```diff
@@ -0,0 +1,166 @@
+package com.lynbrookrobotics.potassium.streams
+
+import com.lynbrookrobotics.potassium.clock.Clock
+import squants.time.Time
+
+import scala.collection.mutable
+import scala.ref.WeakReference
+
+abstract class Stream[T] {
+  val expectedPeriodicity: ExpectedPeriodicity
+  private[this] val listeners = mutable.Queue.empty[T => Unit]
+
+  protected def publishValue(value: T): Unit = {
+    listeners.foreach(_(value))
+    // TODO: more stuff maybe
+  }
+
+  /**
+    * Merges two streams, with a value published from the resulting stream
+    * whenever the parent streams have both published a new value
+    * @param other the stream to merge with
+    * @tparam O the type of values from the other stream
+    * @return a stream with the values from both streams brought together
+    */
+  def zip[O](other: Stream[O]): Stream[(T, O)] = {
+    val ret = new ZippedStream[T, O](expectedPeriodicity, other.expectedPeriodicity)
+    val ptr = WeakReference(ret)
+
+    var aCancel: Cancel = null
+    aCancel = this.foreach { a =>
```

Comment:
Calling `foreach` adds a listener and returns a function to call when we want to cancel the subscription.